crates/api/src/constraints/api.rs

```diff
@@ -0,0 +1,461 @@
+use axum::{
+    body::{to_bytes, Body},
+    http::{Request, StatusCode},
+    Extension,
+};
+use ethereum_consensus::{deneb::Slot, ssz};
+use helix_common::{
+    api::constraints_api::{
+        SignableBLS, SignedDelegation, SignedRevocation, DELEGATION_ACTION,
+        MAX_CONSTRAINTS_PER_SLOT, REVOCATION_ACTION,
+    },
+    bellatrix::List,
+    chain_info::ChainInfo,
+    proofs::{ConstraintsMessage, SignedConstraints, SignedConstraintsWithProofData},
+    ConstraintSubmissionTrace,
+};
+use helix_database::DatabaseService;
+use helix_datastore::Auctioneer;
+use helix_utils::signing::{verify_signed_message, COMMIT_BOOST_DOMAIN};
+use std::{
+    collections::HashSet,
+    sync::Arc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+use tokio::{sync::broadcast, time::Instant};
+use tracing::{error, info, trace, warn};
+use uuid::Uuid;
+
+use crate::constraints::error::ConstraintsApiError;
+
+use super::error::Conflict;
+
+// This is the maximum length (randomly chosen) of a request body in bytes.
+pub(crate) const MAX_REQUEST_LENGTH: usize = 1024 * 1024 * 5;
+
+#[derive(Clone)]
+pub struct ConstraintsApi<A, DB>
+where
+    A: Auctioneer + 'static,
+    DB: DatabaseService + 'static,
+{
+    auctioneer: Arc<A>,
+    db: Arc<DB>,
+    chain_info: Arc<ChainInfo>,
+
+    constraints_handle: ConstraintsHandle,
+}
+
+#[derive(Clone)]
+pub struct ConstraintsHandle {
+    pub(crate) constraints_tx: broadcast::Sender<SignedConstraints>,
+}
+
+impl ConstraintsHandle {
+    pub fn send_constraints(&self, constraints: SignedConstraints) {
+        if self.constraints_tx.send(constraints).is_err() {
+            error!("Failed to send constraints to the constraints channel");
+        }
+    }
+}
+
+impl<A, DB> ConstraintsApi<A, DB>
+where
+    A: Auctioneer + 'static,
+    DB: DatabaseService + 'static,
+{
+    pub fn new(
+        auctioneer: Arc<A>,
+        db: Arc<DB>,
+        chain_info: Arc<ChainInfo>,
+        constraints_handle: ConstraintsHandle,
+    ) -> Self {
+        Self { auctioneer, db, chain_info, constraints_handle }
+    }
+
+    /// Handles the submission of batch of signed constraints.
+    ///
+    /// Implements this API: <https://docs.boltprotocol.xyz/api/builder#constraints>
```

Comment:
broken link, update to https://docs.boltprotocol.xyz/technical-docs/api/builder#constraints